data-pipeline-flink/device-profile-updater/src/main/scala/org/sunbird/dp/deviceprofile/functions/DeviceProfileUpdaterFunction.scala

```diff
@@ -0,0 +1,198 @@
+    val lastUpdatedDate: Long = deviceData.get(config.apiLastUpdatedOn).asInstanceOf[Number].longValue()
+    val parsedKeys = new util.ArrayList[String](util.Arrays.asList("first_access", config.apiLastUpdatedOn))
+    deviceData.keySet.removeAll(parsedKeys)
+    val columns = String.join(",", deviceData.keySet())
+    val values = StringUtils.repeat("?,", deviceData.values.size - 1);
+    /**
+     * Inserting data into postgres
+     */
+    val postgresQuery = String.format("INSERT INTO %s (api_last_updated_on,updated_date,%s) VALUES(?,?,%s?) ON CONFLICT(device_id) DO UPDATE SET (api_last_updated_on,updated_date,%s)=(?,?,%s?);", config.postgresTable, columns, values, columns, values)
+
+    val preparedStatement = postgresConnect.getConnection.prepareStatement(postgresQuery)
+    preparedStatement.setTimestamp(1, new Timestamp(lastUpdatedDate)) // Adding api_last_updated_on as timestamp to index 1 of preparestatement
+
+    preparedStatement.setTimestamp(deviceData.values.size + 3, new Timestamp(lastUpdatedDate)) // Adding api_last_updated_on as timestamp to 3rd index after the map size(for on conflict value)
+
+    preparedStatement.setTimestamp(2, new Timestamp(System.currentTimeMillis)) // Adding updated_date as timestamp to index 2 of preparestatement
+
+    preparedStatement.setTimestamp(deviceData.values.size + 4, new Timestamp(System.currentTimeMillis)) // Adding updated_date as timestamp to 4th index after the map size(for on conflict value)
+
+    setPrepareStatement(preparedStatement, 2, deviceData.asInstanceOf[util.Map[String, AnyRef]]) // Adding map values to preparestatement from index after the api_last_updated_on and updated_on
+
+    setPrepareStatement(preparedStatement, deviceData.values().size() + 4, deviceData.asInstanceOf[util.Map[String, AnyRef]])
+
+    preparedStatement.executeUpdate
+    preparedStatement.close()
+    // Update first_access column if the first_access is null
+    val updateFirstAccessQuery = String.format("UPDATE %s SET first_access = '%s' WHERE device_id = '%s' AND first_access IS NULL", config.postgresTable, new Timestamp(firstAccess).toString, deviceId)
+    postgresConnect.execute(updateFirstAccessQuery)
+
+    if (null != deviceData.get(config.userDeclaredState)) {
+      // Updating user_declared_on column
+      val updateUserDeclaredOnQuery = String.format("UPDATE %s SET user_declared_on = '%s' WHERE device_id = '%s' AND user_declared_on IS NULL", config.postgresTable, new Timestamp(lastUpdatedDate).toString, deviceId)
+      postgresConnect.execute(updateUserDeclaredOnQuery)
+    }
+    logger.info(s"Device ID: $deviceId Updated the postgres device profile table")
+  }
+
+  private def addDeviceDataToCache(deviceId: String, deviceProfile: DeviceProfile): util.Map[String, String] = {
+    val deviceMap = deviceProfile.toMap(config)
+    deviceMap.values.removeAll(Collections.singleton(""))
+    deviceMap.values.removeAll(Collections.singleton("{}"))
```

Comment:
Aren't this redundant steps?

Reply:
The reason is actually we are only updating the only missing fields into Redis.
If the missing field value is empty `string/object` then we don't want to add those fields into Redis.